main/autogen/autoloader.cc

```diff
@@ -400,17 +407,84 @@ void AutoloadWriter::write(const core::GlobalState &gs, const AutoloaderConfig &
                            UnorderedSet<std::string> &toDelete, const DefTree &node) {
     string name = node.root() ? "root" : node.name().show(gs);
     string filePath = join(path, fmt::format("{}.rb", name));
-    FileOps::writeIfDifferent(filePath, node.renderAutoloadSrc(gs, alCfg));
+    if (!alCfg.packagedAutoloader || !node.root()) {
+        FileOps::writeIfDifferent(filePath, node.renderAutoloadSrc(gs, alCfg));
+    }
     toDelete.erase(filePath);
     if (!node.children.empty()) {
         auto subdir = join(path, node.root() ? "" : name);
         if (!node.root() && !FileOps::dirExists(subdir)) {
             FileOps::createDir(subdir);
         }
         for (auto &[_, child] : node.children) {
-            write(gs, alCfg, subdir, toDelete, *child);
+            if (alCfg.packagedAutoloader && node.root()) {
+                // in a packaged context, we want to make sure that these constants are also put in their packages
+                ENFORCE(child->qname.package);
+                auto namespaceName = child->qname.package->show(gs);
+                // this package name will include the suffix _Package on the end, and we want to remove that
+                constexpr int suffixLen = "_Package"sv.length();
+                ENFORCE(namespaceName.size() > suffixLen);
+                string packageName = namespaceName.substr(0, namespaceName.size() - suffixLen);
+                auto pkgSubdir = join(subdir, packageName);
+                FileOps::ensureDir(pkgSubdir);
+                write(gs, alCfg, pkgSubdir, toDelete, *child);
+            } else {
+                write(gs, alCfg, subdir, toDelete, *child);
+            }
         }
     }
 }
 
+string renderPackageAutoloadSrc(const core::GlobalState &gs, const AutoloaderConfig &alCfg, const Package &pkg,
```

Comment:
Nit: you can put this into an anonymous namespace so it is only internally linked.